### PR TITLE
improve log_softmax multi-core performance

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76279
* #74899
* #74898
* #74897

Differential Revision: [D36337195](https://our.internmc.facebook.com/intern/diff/D36337195)

This PR is to improve `logsoftmax` multi-core performance by decoupling `grain_size` and `chunk_size`.

This original implementation has a minimum `grain_size` of `chunk_size` so that log can use vectorized implementation. This limits the parallelism for most of the input shapes in transformers. For example, the shape `[512, 28996]` would use only use 2 cores at most. We can decouple `grain_size` and `chunk_size`, the vec::map<> will still ensure to use vectorized implementation (it will add padding if necessary and do a vectorized log).

Benchmark data on Xeon 6248, 20 cores per socket, dual socket, 2.50GHz:

### cpu performance on single socket (20 cores):
```
### before
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([512, 28996]) torch.float32 time: 14.025 ms
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([511, 250002]) torch.float32 time: 119.716 ms

### after
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([512, 28996]) torch.float32 time: 2.715 ms
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([511, 250002]) torch.float32 time: 26.520 ms
```

### cpu performance on 4 cores run
```
### before
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([512, 28996]) torch.float32 time: 13.166 ms
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([511, 250002]) torch.float32 time: 119.698 ms

### after
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([512, 28996]) torch.float32 time: 6.760 ms
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([511, 250002]) torch.float32 time: 62.041 ms
```

### cpu performance on 1 core run
```
### before
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([512, 28996]) torch.float32 time: 25.599 ms
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([511, 250002]) torch.float32 time: 232.332 ms

### after
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([512, 28996]) torch.float32 time: 25.764 ms
<class 'torch.nn.modules.activation.LogSoftmax'> torch.Size([511, 250002]) torch.float32 time: 232.528 ms
```

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by mingfeima** [comment]
TODO:
- [ ] experiment with implementation with parallel on inner dimension of logsoftmax.

With commonly used shapes in logsoftmax from Transformer models, the lastdim is quite large (20k~300k). And the kernel of logsoftmax requires to read the input for 3 times, so it is possibly to be swapped out of cache.

The cuda kernel utilized different block setup rules for different input shapes, which will go parallel on lastdim in case it is big enough so as to improve cache hit rate for the input data.

Try to implement similar kernel for CPU (need to plant it into 3rd party, e.g. fbgemm, since the kernel requires sync inside omp loops and currently aten won't compile it)

**Comment by frank-wei** [comment]
@frank-wei has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36337195).

**Comment by frank-wei** [comment]
@frank-wei has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36337195).

**Comment by frank-wei** [comment]
Pls check failed test cases.

**Comment by sanchitintel** [comment]
Hi @mingfeima, can we add this change in #85398? #85398's merging had failed earlier but I made more changes & @peterbell10 approved it y'day. Thanks!

**Comment by mingfeima** [comment]
> Hi @mingfeima, can we add this change in #85398? #85398's merging had failed earlier but I made more changes & @peterbell10 approved it y'day. Thanks!

Oh, cool. I will close this one!
